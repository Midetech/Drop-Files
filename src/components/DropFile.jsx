import React, { Component } from 'react';
import uuid from 'uuid';
import PropTypes from 'prop-types';
import './styles/styles.scss';
import './font-awesome-4.7.0/scss/font-awesome.scss';

/**
 * @description Drop File Component
 * 
 * @class DropFile
 * @extends {Component}
 */
class DropFile extends Component{
  constructor(props){
    super(props);
    this.state = {
      fileExtensionsRegex: /\.dmg|\.apk|\.docx|\.m4a/,
      supportedFormats: {
        video: /video/,
        image: /image/,
        audio: /audio/
      }
    }
    this.onChange = this.onChange.bind(this);
    this.dragAndDropFile = this.dragAndDropFile.bind(this);
    this.removeFile = this.removeFile.bind(this);
    this.renderFilePreviewer = this.renderFilePreviewer.bind(this);
  }

  /**
   * @description change state
   * 
   * @method onChange
   * 
   * @param {Object} event
   * 
   * @memberof DropFile
   */
  onChange(event){
    event.preventDefault();
    const { that } = this.props;
    const { fileExtensionsRegex } = this.state
    const { name } = event.target;
    const { files } = event.target;
    let arrayOfFiles = Array.from(files);
    arrayOfFiles = arrayOfFiles
    .filter(file => !fileExtensionsRegex.test(file.name))
    that.setState(
      { [name]: [...that.state.files, ...arrayOfFiles], showFilePreviewer: true }
    )
  }

  dragAndDropFile(event){
    event.preventDefault();
    event.stopPropagation();
    const { that } = this.props;
    const data = event.dataTransfer;
    const { fileExtensionsRegex } = this.state
    const { files } = data
    if(files.length > 0){
      let arrayOfFiles = Array.from(files);
      arrayOfFiles = arrayOfFiles.filter(file => {
        return !fileExtensionsRegex.test(file.name)
      })  
      that.setState(
        { files: [...that.state.files, ...arrayOfFiles], showFilePreviewer: true }
      )
    }


  }
  /**
   * @description remove file from state
   * 
   * @method removeFile
   * 
   * @param {Object} event
   * 
   * @memberof DropFile
   */
  removeFile(event){
    event.preventDefault();
    const { that } = this.props;
    const fileIndex = Number(event.target.dataset.id);
    const files = that.state.files.filter((files, index) => index !== fileIndex);
    that.setState({ files });
  }
    

  /**
   * @description render file previewer
   * 
   * @method render
   * 
   * @returns {Jsx} Jsx
   * @memberof DropFile
   */
  renderFilePreviewer(){
    const { that } = this.props;
    const { 
      fileExtensionsRegex,
      supportedFormats: { image, video, audio }
    } = this.state
    window.URL = window.URL || window.webkitURL;
      return (
        that.state.files.length > 0 ?
        <div id="filePreviewer"
        display="showPreviewer"
      >
        { 
          that.state.files.map((file, index) =>
            {
            if(image.test(file.type)){
              return (
                <a href={window.URL.createObjectURL(file)}
                className="imageThumbnails"
                key={uuid()}
                >
                <img src={window.URL.createObjectURL(file)} data-id={index}
                />
                <span className="fa fa-trash" data-id={index}
                onClick={this.removeFile}
                />
                </a>
              )
            } else if(audio.test(file.type)){
              return (
                <audio controls key={uuid()}>
                  <source src={window.URL.createObjectURL(file)}/>
                </audio>
              )
            } else if(video.test(file.type)){
              return (
                <video className="imageThumbnails" key={uuid()}
                  preload controls
                >
                  <source src={window.URL.createObjectURL(file)} />
                </video>
              )
            }
              return(
                <div className="fileThumbnails"
                  key={uuid()}
                >
                  <object data={window.URL.createObjectURL(file)}
                  type={file.type}
                  >Not Supported</object>
                  <span className="fa fa-trash" data-id={index} 
                    onClick={this.removeFile}
                  />
                </div>
              )
        })
        }
      </div>
      : ''
      )
  };

  /**
   * 
   * 
   * @returns {Jsx} Jsx
   * @memberof DropFile
   */
  render(){
      return (
      <main>
        <div>
            { this.renderFilePreviewer() }
            <label id="selectFileContainer" htmlFor='file'
              onDragEnter={this.dragAndDropFile}
              onDragOver={this.dragAndDropFile}
              onDrop={this.dragAndDropFile}
            >
                <span>Drag & Drop files</span>
                <span className="fa fa-camera"></span>
                <span>Or Click</span>
            </label>
            <input type="file" style={{display:"none"}} name='files'
              multiple
              id="file" onChange={this.onChange}
            />
        </div>
      </main>
      );
    }
}

const propTypes = {
  that: PropTypes.object.isRequired
}

DropFile.propTypes = propTypes;

export default DropFile;