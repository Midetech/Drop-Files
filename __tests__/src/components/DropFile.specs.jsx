import React from 'react';
import { shallow } from 'enzyme';
import DropFile from '../../../index';

const event = {
  preventDefault: jest.fn(),
  stopPropagation: jest.fn(),
  dataTransfer: {
    files: ['fileone']
  },
  target: {
    id: 'files',
    name: 'files',
    dataset: {
      id: 'files'
    },
    files: ['one']
  }
}
const props = {
  that: {
    state: {
      files: [{}]
    },
    setState: jest.fn()
  }
}

describe('# Drop File', () => {

  it('should render successfully', (done) => {
    const wrapper = shallow(<DropFile
      {...props}
      />);
    expect(wrapper).toBeDefined();
    expect(wrapper.getElement().type).toBe('main');
    expect(wrapper).toMatchSnapshot()
    done();
  });

  it('should click on input', (done) => {
    const onChangeSpy = jest.spyOn(DropFile.prototype, 'onChange')
    const wrapper = shallow(<DropFile
      {...props}
      />);
    wrapper.find('#file').simulate('change', event)
    expect(onChangeSpy).toHaveBeenCalled();
    done();
  });

  it('should drag and drop file', (done) => {
    const dragAndDropFileSpy = jest.spyOn(DropFile.prototype, 'dragAndDropFile')
    const wrapper = shallow(<DropFile
      {...props}
      />);
    wrapper.find('#selectFileContainer').simulate('dragenter', event)
    wrapper.find('#selectFileContainer').simulate('dragover', event)
    wrapper.find('#selectFileContainer').simulate('drop', event)
    expect(dragAndDropFileSpy).toHaveBeenCalled();
    done();
  });

  it('should remove file', (done) => {
    const removeFileSpy = jest.spyOn(DropFile.prototype, 'removeFile')
    const wrapper = shallow(<DropFile
      {...props}
      />);
    wrapper.find('[data-id=0]').simulate('click', event)
    expect(removeFileSpy).toHaveBeenCalled();
    done();
  });
});