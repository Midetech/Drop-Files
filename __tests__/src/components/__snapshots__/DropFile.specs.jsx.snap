// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Drop File should render successfully 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DropFile
    that={
      Object {
        "setState": [MockFunction],
        "state": Object {
          "files": Array [
            Object {},
          ],
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <div
          display="showPreviewer"
          id="filePreviewer"
        >
          <div
            className="fileThumbnails"
          >
            <object
              data={undefined}
              type={undefined}
            >
              Not Supported
            </object>
            <span
              className="fa fa-trash"
              data-id={0}
              onClick={[Function]}
            />
          </div>
        </div>
        <label
          htmlFor="file"
          id="selectFileContainer"
          onDragEnter={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        >
          <span>
            Drag & Drop files
          </span>
          <span
            className="fa fa-camera"
          />
          <span>
            Or Click
          </span>
        </label>
        <input
          id="file"
          multiple={true}
          name="files"
          onChange={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          type="file"
        />
      </div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            display="showPreviewer"
            id="filePreviewer"
          >
            <div
              className="fileThumbnails"
            >
              <object
                data={undefined}
                type={undefined}
              >
                Not Supported
              </object>
              <span
                className="fa fa-trash"
                data-id={0}
                onClick={[Function]}
              />
            </div>
          </div>,
          <label
            htmlFor="file"
            id="selectFileContainer"
            onDragEnter={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          >
            <span>
              Drag & Drop files
            </span>
            <span
              className="fa fa-camera"
            />
            <span>
              Or Click
            </span>
          </label>,
          <input
            id="file"
            multiple={true}
            name="files"
            onChange={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            type="file"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="fileThumbnails"
              >
                <object
                  data={undefined}
                  type={undefined}
                >
                  Not Supported
                </object>
                <span
                  className="fa fa-trash"
                  data-id={0}
                  onClick={[Function]}
                />
              </div>,
            ],
            "display": "showPreviewer",
            "id": "filePreviewer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "5144db63-294a-4f92-870c-2f0100d0077e",
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <object
                    data={undefined}
                    type={undefined}
                  >
                    Not Supported
                  </object>,
                  <span
                    className="fa fa-trash"
                    data-id={0}
                    onClick={[Function]}
                  />,
                ],
                "className": "fileThumbnails",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Not Supported",
                    "data": undefined,
                    "type": undefined,
                  },
                  "ref": null,
                  "rendered": "Not Supported",
                  "type": "object",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "fa fa-trash",
                    "data-id": 0,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span>
                Drag & Drop files
              </span>,
              <span
                className="fa fa-camera"
              />,
              <span>
                Or Click
              </span>,
            ],
            "htmlFor": "file",
            "id": "selectFileContainer",
            "onDragEnter": [Function],
            "onDragOver": [Function],
            "onDrop": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Drag & Drop files",
              },
              "ref": null,
              "rendered": "Drag & Drop files",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "fa fa-camera",
              },
              "ref": null,
              "rendered": null,
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Or Click",
              },
              "ref": null,
              "rendered": "Or Click",
              "type": "span",
            },
          ],
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "id": "file",
            "multiple": true,
            "name": "files",
            "onChange": [Function],
            "style": Object {
              "display": "none",
            },
            "type": "file",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
    "type": "main",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <div
            display="showPreviewer"
            id="filePreviewer"
          >
            <div
              className="fileThumbnails"
            >
              <object
                data={undefined}
                type={undefined}
              >
                Not Supported
              </object>
              <span
                className="fa fa-trash"
                data-id={0}
                onClick={[Function]}
              />
            </div>
          </div>
          <label
            htmlFor="file"
            id="selectFileContainer"
            onDragEnter={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          >
            <span>
              Drag & Drop files
            </span>
            <span
              className="fa fa-camera"
            />
            <span>
              Or Click
            </span>
          </label>
          <input
            id="file"
            multiple={true}
            name="files"
            onChange={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            type="file"
          />
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              display="showPreviewer"
              id="filePreviewer"
            >
              <div
                className="fileThumbnails"
              >
                <object
                  data={undefined}
                  type={undefined}
                >
                  Not Supported
                </object>
                <span
                  className="fa fa-trash"
                  data-id={0}
                  onClick={[Function]}
                />
              </div>
            </div>,
            <label
              htmlFor="file"
              id="selectFileContainer"
              onDragEnter={[Function]}
              onDragOver={[Function]}
              onDrop={[Function]}
            >
              <span>
                Drag & Drop files
              </span>
              <span
                className="fa fa-camera"
              />
              <span>
                Or Click
              </span>
            </label>,
            <input
              id="file"
              multiple={true}
              name="files"
              onChange={[Function]}
              style={
                Object {
                  "display": "none",
                }
              }
              type="file"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  className="fileThumbnails"
                >
                  <object
                    data={undefined}
                    type={undefined}
                  >
                    Not Supported
                  </object>
                  <span
                    className="fa fa-trash"
                    data-id={0}
                    onClick={[Function]}
                  />
                </div>,
              ],
              "display": "showPreviewer",
              "id": "filePreviewer",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "5144db63-294a-4f92-870c-2f0100d0077e",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <object
                      data={undefined}
                      type={undefined}
                    >
                      Not Supported
                    </object>,
                    <span
                      className="fa fa-trash"
                      data-id={0}
                      onClick={[Function]}
                    />,
                  ],
                  "className": "fileThumbnails",
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Not Supported",
                      "data": undefined,
                      "type": undefined,
                    },
                    "ref": null,
                    "rendered": "Not Supported",
                    "type": "object",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "className": "fa fa-trash",
                      "data-id": 0,
                      "onClick": [Function],
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "span",
                  },
                ],
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <span>
                  Drag & Drop files
                </span>,
                <span
                  className="fa fa-camera"
                />,
                <span>
                  Or Click
                </span>,
              ],
              "htmlFor": "file",
              "id": "selectFileContainer",
              "onDragEnter": [Function],
              "onDragOver": [Function],
              "onDrop": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Drag & Drop files",
                },
                "ref": null,
                "rendered": "Drag & Drop files",
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "fa fa-camera",
                },
                "ref": null,
                "rendered": null,
                "type": "span",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Or Click",
                },
                "ref": null,
                "rendered": "Or Click",
                "type": "span",
              },
            ],
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "id": "file",
              "multiple": true,
              "name": "files",
              "onChange": [Function],
              "style": Object {
                "display": "none",
              },
              "type": "file",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "div",
      },
      "type": "main",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
